src/main/java/com/example/gradecalculator/service/UserService.java

```diff
@@ -11,6 +11,12 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.stereotype.Service;
+import org.springframework.validation.BindingResult;
+import org.springframework.validation.ObjectError;
+
+import java.util.ArrayList;
+
+import static com.example.gradecalculator.controller.UserController.validate;
```

Comment:
this is not good, references should not be circular <br>
UserService <-> UserController<br>
move the needed static method into the service